### PR TITLE
Add support for accessibility for nightwatch

### PR DESCRIPTION
**Pull request by karanshah-browserstack**

### CODE REVIEW

File: src/accessibilityAutomation.js [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+            let currentURL = await browser.getCurrentUrl();
+
+            let url;
+            try {
+              url = new URL(currentURL);
+              pageOpen = true;
+            } catch (e) {
+              pageOpen = false;
+            }
+            // pageOpen = url.protocol === 'http:' || url.protocol === 'https:';
```
Comment (rev-doshi) [inline src/accessibilityAutomation.js:360]: Why is this commented out? Its part of the logic for pageOpen AFAIR

File: src/accessibilityAutomation.js [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+          helper.getBuildName(this._settings, this._bstackOptions, fromProduct) ||
+          path.basename(path.resolve(process.cwd())),
+        startTime: new Date().toISOString(),
+        description: accessibilityOptions.buildDescription || '',
+        source: {
+          frameworkName: helper.getFrameworkName(this._testRunner),
+          frameworkVersion: helper.getPackageVersion('nightwatch'),
+          sdkVersion: helper.getAgentVersion(),
+        },
+        settings: this.filterAccessibilityOptions(accessibilityOptions),
```
Comment (rev-doshi) [inline src/accessibilityAutomation.js:75]: Here we can send the accessibilityOptions itself, don't need the filtered Options

File: src/accessibilityAutomation.js (hunk 1 above)
Comment (karanshah-browserstack) [reply src/accessibilityAutomation.js]: Done

**Review by rev-doshi — APPROVED** [review]

File: nightwatch/globals.js [hunk 3]
```diff
@@ -86,6 +88,20 @@ module.exports = {
       Logger.error(`Could not configure or launch test observability - ${error}`);
     }
 
+    try{
```
Comment (07souravkunda) [inline nightwatch/globals.js:91]: lint: space

File: src/accessibilityAutomation.js [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+    if (this._settings.accessibility) {
+      process.env.BROWSERSTACK_ACCESSIBILITY =
+        String(this._settings.accessibility).toLowerCase() == 'true';
+    }
+    if (process.argv.includes('--disable-accessibility')) {
+      process.env.BROWSERSTACK_ACCESSIBILITY = false;
+      return;
+    }
+    process.env.BROWSERSTACK_INFRA = true;
+    if(settings && settings.webdriver && settings.webdriver.host && settings.webdriver.host.indexOf('browserstack') == -1){
```
Comment (07souravkunda) [inline src/accessibilityAutomation.js:20]: create key extractor utility

Comment (07souravkunda) [reply src/accessibilityAutomation.js:20]: ===

File: src/accessibilityAutomation.js [hunk 5]
```diff
@@ -0,0 +1,517 @@
+const path = require('path');
+const helper = require('./utils/helper');
+const { makeRequest } = require('./utils/requestHelper');
+const Logger = require('./utils/logger');
+const { ACCESSIBILITY_URL } = require('./utils/constants');
+
+class AccessibilityAutomation {
+  configure(settings = {}) {
+    this._settings = settings['@nightwatch/browserstack'] || {};
+
```
Comment (07souravkunda) [inline src/accessibilityAutomation.js:10]: Initialise all class methods in constructor

File: src/accessibilityAutomation.js [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+        if (error.message == 'Invalid configuration passed.') {
+          Logger.error(
+            `Exception while creating test run for BrowserStack Accessibility Automation: ${
+              error.message || error.stack
+            }`
+          );
+          for (const errorkey of error.errors) {
+            Logger.error(errorkey.message);
+          }
+          process.env.BROWSERSTACK_ACCESSIBILITY = 'false';
```
Comment (07souravkunda) [inline src/accessibilityAutomation.js:115]: don't set env vars inside this class, set them as class properties. Set all envs in globals.js|before at one place

File: src/accessibilityAutomation.js [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+        }
+      }
+      return [null, null];
+    }
+  }
+
+  async stopAccessibilityTestRun() {
+    if (
+      helper.isUndefined(process.env.BS_A11Y_JWT) ||
+      typeof process.env.BS_A11Y_JWT !== 'string'
```
Comment (07souravkunda) [inline src/accessibilityAutomation.js:131]: use class property

File: src/accessibilityAutomation.js [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+        process.env.BS_A11Y_JWT !== 'null';
+      return isBrowserstackAccessibilityEnabled && hasA11yJwtToken;
+    } catch (error) {
+      Logger.debug(`Exception in verifying the Accessibility session with error : ${error}`);
+    }
+    return false;
+  }
+
+  shouldScanTestForAccessibility(test) {
+    if (process.env.BROWSERSTACK_ACCESSIBILITY_OPTIONS == null) {
```
Comment (07souravkunda) [inline src/accessibilityAutomation.js:263]: ===

**Review by 07souravkunda — CHANGES_REQUESTED** [review]

File: src/accessibilityAutomation.js [hunk 9]
```diff
@@ -0,0 +1,517 @@
...
+      const excludeTags = Array.isArray(accessibilityConfig.excludeTagsInTestingScope)
+        ? accessibilityConfig.excludeTagsInTestingScope
+        : [];
+
+      const fullTestName = test.module;
+      const excluded = excludeTags.some((exclude) => fullTestName.includes(exclude));
+      const included =
+        includeTags.length === 0 || includeTags.some((include) => fullTestName.includes(include));
+
+      return !excluded && included;
```
Comment (07souravkunda) [inline src/accessibilityAutomation.js:280]: if test is not included or excluded it will return false?

**Review by 07souravkunda — CHANGES_REQUESTED** [review]

**Review by 07souravkunda — APPROVED** [review]

File: nightwatch/globals.js [hunk 10]
```diff
@@ -105,6 +128,25 @@ module.exports = {
...
+        Logger.error(`Exception in stop accessibility test run: ${error}`);
+      }
+      
+    }
+  },
+
+  async beforeEach(settings) {
+    browser.getAccessibilityResults = () =>  { return accessibilityAutomation.getAccessibilityResults(); }
+    browser.getAccessibilityResultsSummary = () => { return accessibilityAutomation.getAccessibilityResultsSummary() };
+    // await accessibilityAutomation.beforeEachExecution(browser);
```
Comment (07souravkunda) [inline nightwatch/globals.js:144]: remove comments

File: nightwatch/globals.js [hunk 11]
```diff
@@ -105,6 +128,25 @@ module.exports = {
...
+  },
+  
+  // This will be run after each test suite is finished
+  async afterEach(settings) {
+    // await accessibilityAutomation.afterEachExecution(browser);
```
Comment (07souravkunda) [inline nightwatch/globals.js:149]: comments

**Review by 07souravkunda — CHANGES_REQUESTED** [review]

**Comment by AutomatedTester** [comment]
Before we can merge this we need to have a green ci. can someone sort that?

**Comment by AutomatedTester** [comment]
@karanshah-browserstack @07souravkunda can someone look at the CI failures?

**Comment by karanshah-browserstack** [comment]
@AutomatedTester Checking

**Comment by karanshah-browserstack** [comment]
@AutomatedTester Fixed this.

**Comment by karanshah-browserstack** [comment]
RUN_TESTS

**Review by kamal-kaur04 — APPROVED** [review]

**Review by 07souravkunda — APPROVED** [review]

**Pull request merged**